parl/layers/layer_wrappers.py

```diff
@@ -73,31 +55,149 @@ def check_or_replace_name(name, attr):
+            target_var.set(src_var, place)
+
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        ## __new__ won't init the class, we need to do that ourselves
+        copied = cls.__new__(cls)
+        ## record in the memo that self has been copied to avoid recursive copying
+        memo[id(self)] = copied
+
+        ## first copy all content
+        for k, v in self.__dict__.items():
+            setattr(copied, k, deepcopy(v, memo))
+
+        ## then we need to create new para names for self.param_attr and self.bias_attr
+        def create_new_para_name(attr):
+            if attr:
+                assert attr.name, "attr should have a name already!"
+                ## remove the last number id but keep the name key
+                name_key = "_".join(attr.name.split("_")[:-1])
+                attr.name = unique_name.generate(name_key)
+
+        create_new_para_name(copied.param_attr)
+        create_new_para_name(copied.bias_attr)
+        ## We require the user to sync the parameter values later, because
+        ## this deepcopy is supposed to be called only before the startup
+        ## program. This function will cause the computation graph change, so
+        ## it cannot be called during the execution.
+        return copied
+
+    @property
+    def param_name(self):
+        if self.param_attr:
+            return self.param_attr.name
+        else:
+            return None
+
+    @property
+    def bias_name(self):
+        if self.bias_attr:
+            return self.bias_attr.name
+        else:
+            return None
+
+
+class Network(object):
+    """
+    A Network is an unordered set of LayerFuncs or Networks.
+    """
+
+    def sync_paras_to(self, target_net, gpu_id):
+        assert not target_net is self, "cannot copy between identical networks"
+        assert isinstance(target_net, Network)
+
+        for attr in self.__dict__:
+            if not attr in target_net.__dict__:
+                continue
+            val = getattr(self, attr)
+            target_val = getattr(target_net, attr)
+
+            assert type(val) == type(target_val)
+            ## only these two types of members will be copied
+            ## the others will be ignorped
+            if isinstance(val, Network) or isinstance(val, LayerFunc):
+                val.sync_paras_to(target_val, gpu_id)
+            elif isinstance(val, tuple) or isinstance(val, list) or isinstance(
+                    val, set):
+                for v, tv in zip(val, target_val):
+                    v.sync_paras_to(tv, gpu_id)
+            elif isinstance(val, dict):
+                for k in val.keys():
+                    assert k in target_val
+                    val[k].sync_paras_to(target_val[k], gpu_id)
+            else:
+                # for any other type, we do not copy
+                pass
+
+
+def check_caller_name():
+    stack = inspect.stack()
```

Comment:
correctness is still not guaranteed. Because the layer still may not be stored in a Network object even it's on a stack with Network.__init__. Perhaps we should use a similar approach as copy.deepcopy to do sync_paras_to()

Reply:
Added TODO